### ISSUE TITLE
[Python] `pa.array(<pd.Series of structs>)` changes field order to be sorted

### ISSUE THREAD

**Issue by NickCrews** [apache/arrow#40053]
### Describe the bug, including details regarding any error messages, version, and platform.

If I give a column of structs, then pyarrow seems to sort the fields by name. I would expect the field order to be preserved.

```python
import pandas as pd
import pyarrow as pa

print(pa.__version__)
# 15.0.0

s = pd.Series([{"b": 2, "a": 1}, {"b": 4, "a": 3}])
a = pa.array(s)
print(a.type)
# struct<a: int64, b: int64>
```

This was discovered in https://github.com/ibis-project/ibis/issues/8166

### Component(s)

Python

**Comment by jorisvandenbossche** [comment]
Quick guess, but I assume this might stem from a time when Python dicts didn't have an order, and so the keys were sorted to guarantee stable results.

**Comment by jorisvandenbossche** [comment]
Although it seems we iterate over the dict using `PyDict_Next`. The Python docs (https://docs.python.org/3/c-api/dict.html#c.PyDict_Next) are not very clear about the order of the iteration with this function.

**Comment by NickCrews** [comment]
Thanks for the help @jorisvandenbossche !

Are you referring to [this function](https://github.com/apache/arrow/blob/bbe59b35de33a0534fc76c9617aa4746031ce16c/python/pyarrow/src/arrow/python/inference.cc#L628-L660)? Could the shuffling be due to the use of `struct_inferrers_.insert()`, where `struct_inferrers_` is a `std::map<std::string, TypeInferrer>`, so it doesn't preserve order?

**Comment by jorisvandenbossche** [comment]
> Are you referring to [this function](https://github.com/apache/arrow/blob/bbe59b35de33a0534fc76c9617aa4746031ce16c/python/pyarrow/src/arrow/python/inference.cc#L628-L660)?

Yes, and then the `struct_inferrers_` gets converted to the struct type here:

**Comment by jorisvandenbossche** [comment]
> Are you referring to [this function](https://github.com/apache/arrow/blob/bbe59b35de33a0534fc76c9617aa4746031ce16c/python/pyarrow/src/arrow/python/inference.cc#L628-L660)?

Yes, and then the `struct_inferrers_` gets converted to the struct type here: 

https://github.com/apache/arrow/blob/bbe59b35de33a0534fc76c9617aa4746031ce16c/python/pyarrow/src/arrow/python/inference.cc#L668-L676


> Could the shuffling be due to the use of `struct_inferrers_.insert()`, where `struct_inferrers_` is a `std::map<std::string, TypeInferrer>`, so it doesn't preserve order?

Good catch! Indeed, a map will sort the keys. If we want to preserve the order of the dict keys, we will have to use a different data structure (or keep track of the order separately)

**Comment by NickCrews** [comment]
If you point me in the direction of which data structure you think we should use, I can take a stab at this.

I feel like somewhere in the arrow codebase there MUST be somewhere else an ordered map is used, but [I didn't find it](https://github.com/search?q=repo%3Aapache%2Farrow%20ordered%20map&type=code).

I found https://stackoverflow.com/questions/1098175/a-stdmap-that-keep-track-of-the-order-of-insertion. I'm assuming we don't want to bring in any external deps. Could copy-paste https://github.com/Tessil/ordered-map/blob/master/include/tsl/ordered_map.h? Or, its much smaller, but more "one-off", could [implement it with a list and a map.](https://stackoverflow.com/a/1098236/5156887)

**Comment by pitrou** [comment]
Issue resolved by pull request 48813
https://github.com/apache/arrow/pull/48813

**Comment by pitrou** [comment]
@gabrielsimoes Can you comment here so that we assign this issue to you?

**Comment by gabrielsimoes** [comment]
:+1:

**Comment by NickCrews** [comment]
Thank you all!
